packages/openid4vc-holder/src/issuance/utils/IssuerMetadataUtils.ts

```diff
@@ -0,0 +1,219 @@
+          types: foundSupportedCredential.types,
+        })
+      }
+    }
+    // Otherwise it's an inline credential offer that does not reference a supported credential in the issuer metadata
+    else {
+      offeredCredentials.push({
+        inlineCredentialOffer: offeredCredential,
+        offerType: OfferedCredentialType.InlineCredentialOffer,
+        format: getUniformFormat(offeredCredential.format),
+        types: offeredCredential.types,
+      })
+    }
+  }
+
+  return offeredCredentials
+}
+
+export async function getMetadataFromCredentialOffer(
+  credentialOfferPayload: CredentialOfferPayloadV1_0_11,
+  _metadata?: EndpointMetadataResult
+) {
+  const issuer = credentialOfferPayload.credential_issuer
+
+  const metadata =
+    _metadata && _metadata.credentialIssuerMetadata ? _metadata : await MetadataClient.retrieveAllMetadata(issuer)
+  if (!metadata) throw new AriesFrameworkError(`Could not retrieve metadata for OpenId4Vci issuer: ${issuer}`)
+
+  const issuerMetadata = metadata.credentialIssuerMetadata
+  if (!issuerMetadata)
+    throw new AriesFrameworkError(`Could not retrieve issuer metadata for OpenId4Vci issuer: ${issuer}`)
+
+  return { issuer, metadata, issuerMetadata }
+}
+
+export function getSupportedCredentials(opts?: {
+  issuerMetadata: CredentialIssuerMetadata | IssuerMetadataV1_0_08
+  version: OpenId4VCIVersion
+}): CredentialSupported[] {
+  const { issuerMetadata } = opts ?? {}
+  let credentialsSupported: CredentialSupported[]
+  if (!issuerMetadata) {
+    return []
+  }
+  const { version } = opts ?? { version: OpenId4VCIVersion.VER_1_0_11 }
+
+  const usesTransformedCredentialsSupported =
+    version === OpenId4VCIVersion.VER_1_0_08 || !Array.isArray(issuerMetadata.credentials_supported)
+  if (usesTransformedCredentialsSupported) {
+    credentialsSupported = credentialsSupportedV8ToV11((issuerMetadata as IssuerMetadataV1_0_08).credentials_supported)
+  } else {
+    credentialsSupported = (issuerMetadata as CredentialIssuerMetadata).credentials_supported
+  }
+
+  if (credentialsSupported === undefined || credentialsSupported.length === 0) {
+    return []
+  } else {
+    return credentialsSupported
+  }
+}
+
+export function credentialsSupportedV8ToV11(supportedV8: CredentialSupportedTypeV1_0_08): CredentialSupported[] {
+  return Object.entries(supportedV8).flatMap((entry) => {
+    const type = entry[0]
+    const supportedV8 = entry[1]
+    return credentialSupportedV8ToV11(type, supportedV8)
+  })
+}
+
+export function credentialSupportedV8ToV11(
+  key: string,
+  supportedV8: CredentialSupportedV1_0_08
+): CredentialSupported[] {
+  const v8FormatEntries = Object.entries(supportedV8.formats)
+
+  return v8FormatEntries.map((entry) => {
+    const format = entry[0]
+    const credentialSupportBrief = entry[1]
+    if (typeof format !== 'string') {
+      throw Error(`Unknown format received ${JSON.stringify(format)}`)
+    }
+    let credentialSupport: Partial<CredentialSupported> = {}
```

Comment:
We assign a partial type here and later on cast to be fully typed. Can we get rid of the casting?